dm/syncer/validator_checkpoint.go

```diff
@@ -335,34 +305,106 @@ func (c *validatorPersistHelper) persist(tctx *tcontext.Context, loc binlog.Loca
 			args = append(args, []interface{}{
 				c.cfg.SourceID, sourceTable.Schema, sourceTable.Table, r.srcJob.Key,
 				targetTable.Schema, targetTable.Table,
-				srcDataStr, dstDataStr, r.tp, pb.ValidateErrorState_NewErr,
+				string(srcDataStr), string(dstDataStr), r.tp, pb.ValidateErrorState_NewErr,
 			})
 		}
 	}
-	// todo: performance issue when using insert on duplicate? https://asktug.com/t/topic/33147
-	// todo: will this transaction too big? but checkpoint & pending changes should be saved in one tx
-	var err error
+
+	return c.execQueriesWithRetry(tctx, queries, args)
```

Comment:
This function simply executes queries one after another without using transactions. What if it succeeded to synchronize the table status but failed to insert error rows? When the validator recovers from the two tables, inconsistency probably occurs that the statuses of tables might posit an error but we cannot find it from error rows. I don't think this inconsistency is tolerable.

Reply:
> What if it succeeded to synchronize the table status but failed to insert error rows?

after resume, validator will start from previous loc, and find those error rows again

> the statuses of tables might posit an error

status of table can be stop or running, it cannot show whether there's error rows.

Reply:
lgtm